### PR TITLE
backport (v1.12): policy: Fix Deny Precedence Bug

### PR DESCRIPTION
**Pull request by nathanjsweet**
-  ip: Add a helper to assist net.IP -> netip.Addr conversion (from @YutaroHayakawa)
  
   net.IP to netip.Addr conversion using netip.AddrFromSlice has a pitfall
that it interprets an IPv4 net.IP created with net.ParseIP as IPv4
mapped IPv6 address (see code comment for more details). This helper
safely converts net.IP to netip.Addr with preserving an original address
family, but with assumption that given net.IP is not a "real" IPv4
mapped IPv6 address which practical enough for our use cases.

- Add Tests for Deny Precedence Bug:

  Currently, when a broad "Deny" policy is paired with a specific "Unmanaged CIDR" policy, then the "Unmanaged CIDR" policy will still be inserted into the policy map for an endpoint. This results in "Deny" policies not always taking precedence over "Allow" policies. This test confirms the bugs existence.

- Fix Deny Precedence Bug:

  When the policy map state is created CIDRs are now checked against one another to ensure that deny-rules that supersede allow-rules when they should.

  `DenyPreferredInsert` has been refactored to use utility methods that make the complex boolean logic of policy precedence more atomic.

  Add `NetsContainsAny` method to `pkg/ip` to compare cases where one set of networks conatins or is equal to any network in another set.

- endpoint: Add policy.Identity Implementation

  A `policy.Identity` implementation is necessary for the incremental update to the endpoint's policy map that can occur with L7 changes. Valid deny-policy entries may prohibit these L7 changes based on CIDR rules, which are only obtainable by looking up all potentially conflicting policies' labels. Thus `l4.ToMapState` needs access to the identity allocater to lookup "random" identity labels.

- Additional commit from @joestringer to fix concurrent access to the SelectorCache, a bug the original work introduced.

- Additional commit from @nathanjsweet to prevent a potential nil pointer dereference in the endpoint code.

```release-note
policy: Promote Deny Policies from Beta to Stable
```

### CODE REVIEW

**Comment by nathanjsweet** [comment]
/test-backport-1.12

**Comment by nathanjsweet** [comment]
/test-backport-1.12

File: go.mod
Comment (christarazi) [inline go.mod:1]: How come there are changes to dependencies needed for this backport PR?

Comment (nathanjsweet) [reply go.mod:1]: The backport includes Fuzz tests that weren't there before.

File: pkg/policy/mapstate.go [hunk 1]
```diff
@@ -437,23 +547,23 @@ func (keys MapState) insertIfNotExists(key Key, entry MapStateEntry) {
...
+//  3. Change all L3/L4 ALLOW keys on matching port that do not already redirect to
+//     redirect.
+//     - e.g, <ID1>:80=allow,0 -> <ID1>:80=allow,<proxyport>
+//  4. For each L3-only ALLOW key add the corresponding L3/L4 ALLOW redirect if no
+//     L3/L4 key already exists and no L4-only key already exists and one is not added.
+//     - e.g., <ID2>:0=allow,0 -> add <ID2>:80=allow,<proxyport> if <ID2>:80
+//     and 0:80 do not exist
+//  5. If a new L4-only key was added: For each L3-only DENY key add the
+//     corresponding L3/L4 DENY key if no L3/L4 key already exists.
+//     - e.g., <ID3>:0=deny,0 -> add <ID3>:80=deny,0 if <ID3>:80 does not exist
```
Comment (christarazi) [inline pkg/policy/mapstate.go:566]: Unintended diff?

**Review by christarazi — APPROVED** [review]
I reviewed the main commit (93a198d5a827df95da16f4f597a4fb13fb04580a) and diff'd it against the `main` version. The diffs don't seem relevant, so LGTM.

File: pkg/policy/mapstate.go (hunk 1 above)
Comment (nathanjsweet) [reply pkg/policy/mapstate.go:566]: Someone updated this comment section in v1.13 because 1.11 backport also has the weird reformat.

**Review by julianwiedmann — APPROVED** [review]

**Comment by joestringer** [comment]
@nathanjsweet It looks like the docs are not consistent with the release note here, as they still state that deny policies are beta:

https://docs.cilium.io/en/v1.12/policy/language/#id8

Should we update the docs or adjust the release note to state that a bug was fixed with deny precedence?

**Comment by nathanjsweet** [comment]
I guess we can upgrade the docs to say "stable".

**Pull request merged**